semgrep/semgrep/target_manager.py

```diff
@@ -73,6 +84,196 @@ def converted_pipe_targets(targets: Sequence[str]) -> Iterator[Sequence[str]]:
+    size_limit: Set[Path] = Factory(set)
+
+    rule_includes: Dict[str, Set[Path]] = Factory(lambda: defaultdict(set))
+    rule_excludes: Dict[str, Set[Path]] = Factory(lambda: defaultdict(set))
+    rule_size_limit: Dict[str, Set[Path]] = Factory(lambda: defaultdict(set))
+
+    @property
+    def size_limited_paths(self) -> Set[Path]:
+        """
+        All paths that were skipped because they were larger than the max_target_bytes.
+
+        This is needed because sometimes a file is not targeted at first
+        when we initially filter files that are too large,
+        but then a rule includes it.
+        After a rule includes a file that wasn't previously there,
+        and we notice it's larger than the limit,
+        it's recorded in rule_size_limit instead of size_limit.
+
+        But during output, explaining all this would be too much noise,
+        and not particularly helpful since the user would likely just raise the size limit.
+        So we merge these sets here, and usually look at this property.
+        """
+        return {
+            *self.size_limit,
+            *(path for paths in self.rule_size_limit.values() for path in paths),
+        }
+
+    @property
+    def rule_ids_with_skipped_paths(self) -> FrozenSet[str]:
+        """All rule IDs that have skipped paths.
+
+        Note that if a rule ID defines excludes/includes,
+        but they didn't skip any paths,
+        that rule ID will not show up here.
+        """
+        return frozenset(
+            {
+                *(rule_id for rule_id, skips in self.rule_includes.items() if skips),
+                *(rule_id for rule_id, skips in self.rule_excludes.items() if skips),
+            }
+        )
+
+    def __str__(self) -> str:
+        skip_fragments = []
+
+        if self.target_manager.respect_git_ignore:
+            skip_fragments.append("all .gitignored files")
+
+        if self.cli_includes:
+            skip_fragments.append(
+                f"{len(self.cli_includes)} files not matching --include patterns"
+            )
+        if self.cli_excludes:
+            skip_fragments.append(
+                f"{len(self.cli_excludes)} files matching --exclude patterns"
+            )
+        if self.size_limited_paths:
+            skip_fragments.append(
+                f"{len(self.size_limited_paths)} files larger than {self.target_manager.max_target_bytes / 1000 / 1000} MB"
+            )
+        if self.semgrepignored:
+            skip_fragments.append(
+                f"{len(self.semgrepignored)} files matching .semgrepignore patterns"
+            )
+
+        if not skip_fragments:
+            return "no files were skipped"
+
+        message = "skipped: " + ", ".join(skip_fragments)
+
+        message += "\nfor a detailed list of skipped files, run semgrep with the --verbose flag\n"
+        return message
+
+    def yield_verbose_lines(self) -> Iterator[Tuple[Literal[0, 1, 2], str]]:
```

Comment:
Is the intention to have this class also handle the json output?

Reply:
IMO would be cleaner if the thing outputing handled this code. That way we wouldnt need this class to have a pointer to it's parent.